File: server/Features/Constants/Messages.cs
namespace RSPWebAPI.Features.Constants;

public static class Messages
{
  // Common
  public static class Common
  {
    public const string UnexpectedError = "An unexpected error occurred.";
    public const string EndDateMustBeGreaterThanStartDate =
      "The end date should be later than the start date.";
  }

  // Entities
  public static readonly EntityMessages Season = new GenericMessages("Season");
  public static readonly UserMessages User = new();
  public static readonly EnrollmentMessages Enrollment = new();
  public static readonly MentorshipMessages Mentorship = new();
  public static readonly ProblemAttemptMessages ProblemAttempt = new();
  public static readonly MockInterviewMessages MockInterview = new();
  public static readonly SeasonWeekMessages SeasonWeek = new();

  // No CRUD entities
  public static class LeetcodeProblem
  {
    public const string Listed = "Leetcode problems list retrieved successfully.";
    public const string ScrapedSuccessfully = "Leetcode problems list scraped successfully.";
  }

  public static class LeetcodeProblemRecommendation
  {
    public const string Created = "The Leetcode problem recommendation was set up successfully.";
    public const string CreationError =
      "An error occurred while generating the Leetcode recommendation.";
    public const string NoProblemsLeft =
      "No recommendations available because all Leetcode problems have been completed.";
  }

  public static class DummyData
  {
    public const string Generated = "Dummy data were all generated successfully.";
    public const string GenerationError = "An error occurred while generating dummy data.";
  }

  public static class Graduates
  {
    public const string Listed = "Graduates list retrieved successfully.";
    public const string ListError = "An error occurred while retrieving the graduates list.";
  }

  public static class Student
  {
    public const string KickMenteeDoesntExist =
      "No student enrollment was found for the provided details.";
    public const string KickSuccess = "Student successfully removed.";
    public const string KickError = "An error occurred while removing the student.";
    public const string UpdateRolePromotionMenteeDoesntExist =
      "No student enrollment was found for the provided details.";
    public const string UpdateRolePromotionSuccess = "Role promotion updated successfully.";
    public const string UpdateRolePromotionError =
      "An error occurred while updating the role promotion.";
  }
}

// Base
public abstract class EntityMessages
{
  private protected EntityMessages(string entityName, string idSuffix = "Id")
  {
    EntityName = entityName;
    IdSuffix = idSuffix;
  }

  protected string EntityName { get; }
  protected string IdSuffix { get; }
  private string IdLabel => $"{EntityName}{IdSuffix}";

  public string Exists => $"A {EntityName} with this {IdLabel} already exists.";
  public string DoesNotExist => $"No {EntityName} was found with the given {IdLabel}.";
  public string Created => $"{EntityName} created successfully.";
  public string Updated => $"{EntityName} updated successfully.";
  public string Deleted => $"{EntityName} deleted successfully.";
  public string Listed => $"{EntityName} list retrieved successfully.";
  public string CreationError => $"An error occurred while creating the {EntityName}.";
  public string UpdateError => $"An error occurred while updating the {EntityName}.";
  public string DeletionError => $"An error occurred while deleting the {EntityName}.";
  public string ListError => $"An error occurred while retrieving the {EntityName} list.";
}

// Generic entity
internal sealed class GenericMessages : EntityMessages
{
  internal GenericMessages(string name, string idSuffix = "Id")
    : base(name, idSuffix) { }
}

// Concrete entities
public sealed class UserMessages : EntityMessages
{
  internal UserMessages()
    : base("User") { }

  public string EmailExists = "A user with this email is already registered.";
  public string EmailDoesNotExist = "No user was found with the provided email.";
  public string IdDoesNotExist = "No user was found with the given user id.";
  public string Found = "User found successfully.";

  public string NotInSeasonRole(string role) =>
    $"This user is not registered as a {role} for the specified season.";
}

public sealed class EnrollmentMessages : EntityMessages
{
  internal EnrollmentMessages()
    : base("Enrollment") { }

  public string UsersListed = "Enrollment users list retrieved successfully.";
  public string UsersListError = "An error occurred while retrieving the enrollment users list.";
  public string StudentMustHaveRolePromotion = "Students must have a valid role promotion.";
  public string MentorOrCoordinatorMustNotHaveRolePromotion =
    "Mentors and coordinators should not have a role promotion.";
}

public sealed class MentorshipMessages : EntityMessages
{
  internal MentorshipMessages()
    : base("Mentorship") { }

  public string NotPermittedDueToNullMentorOrMentee =
    "The provided mentor or mentee does not exist.";
  public string NotPermittedDueToDifferentSeason =
    "The mentor and mentee belong to different seasons.";
}

public sealed class ProblemAttemptMessages : EntityMessages
{
  internal ProblemAttemptMessages()
    : base("Problem attempt", idSuffix: "") { }

  public string OutOfSeasonDateRange =
    "The problem attempt date falls outside the season's date range.";
}

public sealed class MockInterviewMessages : EntityMessages
{
  internal MockInterviewMessages()
    : base("Mock interview", idSuffix: "") { }

  public string UpdateOnlyInterviewerAllowed =
    "Only interviewers are allowed to update the mock interview.";
  public string DeletionOnlyInterviewerAllowed =
    "Only interviewers are allowed to delete the mock interview.";
  public string InterviewerOrIntervieweeCannotBeFound =
    "The interviewer or interviewee could not be found.";
  public string OutOfSeasonDateRange =
    "The mock interview date falls outside the season's date range.";
  
  // Round review messages
  public string RoundReviewUpdated = "Review status updated successfully";
  public string CustomRoundReviewUpdated = "Custom round review status updated successfully";
  public string LeetcodeRoundReviewUpdated = "Leetcode round review status updated successfully";
  public string OnlyIntervieweeCanUpdateReview = "Only the interviewee can update the review status";
  public string CustomRoundNotFound = "Custom mock interview round not found";
  public string LeetcodeRoundNotFound = "Leetcode mock interview round not found";
  public string RoundUpdateError = "Error updating mock interview round review status";
  public string CustomRoundUpdateError = "Error updating custom mock interview round review status";
  public string LeetcodeRoundUpdateError = "Error updating leetcode mock interview round review status";
}

public sealed class SeasonWeekMessages : EntityMessages
{
  internal SeasonWeekMessages()
    : base("Season week") { }

  public string DatesNotWithinSeasonDates =
    "The season week dates do not align with the season dates.";
}

File: client/src/shared/constants/messages.ts
export const NOTIFICATION_MESSAGES = {
  USER: {
    CREATED: 'User created successfully.',
    UPDATED: 'User updated successfully.',
    DELETED: 'User deleted successfully.',
    SLUG_UPDATED: 'Your profile slug has been updated successfully! Refreshing page...',
    SLUG_UPDATE_FAILED: 'Failed to update slug. Please try again.',
    SLUG_GENERATED: 'A new random slug has been generated for you!',
    SLUG_GENERATE_FAILED: 'Failed to generate random slug. Please try again.',
  },
  SEASON: {
    CREATED: 'Season created successfully.',
    UPDATED: 'Season updated successfully.',
    DELETED: 'Season deleted successfully.',
  },
  SEASON_WEEK: {
    CREATED: 'Season Week created successfully.',
    UPDATED: 'Season Week updated successfully.',
    DELETED: 'Season Week deleted successfully.',
  },
  ENROLLMENT: {
    CREATED: 'Enrollment created successfully.',
    UPDATED: 'Enrollment updated successfully.',
    DELETED: 'Enrollment deleted successfully.',
  },
  MENTORSHIP: {
    CREATED: 'Mentorship created successfully.',
    UPDATED: 'Mentorship updated successfully.',
    DELETED: 'Mentorship deleted successfully.',
  },
  MOCK_INTERVIEW: {
    CREATED: 'Mock interview created successfully.',
    UPDATED: 'Mock interview updated successfully.',
    DELETED: 'Mock interview deleted successfully.',
    REVIEW_UPDATED: 'Review status updated successfully.',
  },
  PROBLEM_ATTEMPT: {
    CREATED: 'Problem attempt created successfully.',
    UPDATED: 'Problem attempt updated successfully.',
    DELETED: 'Problem attempt deleted successfully.',
  },
  MENTEE: {
    DELETED: 'Mentee deleted successfully.',
  },
} as const;

export const CONFIRMATION_MESSAGES = {
  USER: {
    DELETE_TITLE: 'Delete User',
    DELETE_TEXT: 'Are you sure you want to delete this user? This action cannot be undone.',
  },
  SEASON: {
    DELETE_TITLE: 'Delete Season',
    DELETE_TEXT: 'Are you sure you want to delete this season? This action cannot be undone.',
  },
  SEASON_WEEK: {
    DELETE_TITLE: 'Delete Season Week',
    DELETE_TEXT: 'Are you sure you want to delete this season week? This action cannot be undone.',
  },
  ENROLLMENT: {
    DELETE_TITLE: 'Delete Enrollment',
    DELETE_TEXT: 'Are you sure you want to delete this enrollment? This action cannot be undone.',
  },
  MENTORSHIP: {
    DELETE_TITLE: 'Delete Mentorship',
    DELETE_TEXT: 'Are you sure you want to delete this mentorship? This action cannot be undone.',
  },
  MOCK_INTERVIEW: {
    DELETE_TITLE: 'Delete Mock Interview',
    DELETE_TEXT:
      'Are you sure you want to delete this mock interview? This action cannot be undone.',
  },
  PROBLEM_ATTEMPT: {
    DELETE_TITLE: 'Delete Problem Attempt',
    DELETE_TEXT:
      'Are you sure you want to delete this problem attempt? This action cannot be undone.',
  },
  MENTEE: {
    DELETE_TITLE: 'Delete Mentee',
    DELETE_TEXT: 'Are you sure you want to delete this mentee? This action cannot be undone.',
  },
} as const;
